### ISSUE TITLE
[compiler] Rewrite Eta in Eta

### ISSUE THREAD

**Issue by NickSeagull** [typelead/eta#696]
This would use all the changes proposed in the Compiler Extensions proposal, the new prelude, and the new base (after we manage to figure them out)

Another great thing would be if we could change Happy/Alex parsing/lexing for monadic/applicative parsing/lexing. It would make stuff more readable in my opinion.

In general, I find GHC code pretty hard to understand as a newcomer and I like the way [PureScript is structured](https://github.com/purescript/purescript/blob/master/src/Language/PureScript/Parser/Lexer.hs) for example.

Another advantage of this is that one could compile newer versions of Eta using a binary one.

**Comment by jneira** [comment]
This is a big one! i think it is in some way a duplicate of #423 
Anyway, imo it would not have to be a high priority goal, except as a means of getting something else (like repl support)

**Comment by rahulmutt** [comment]
:+1: for swapping out Happy/Alex with monadic/applicative lexing. PureScript is a great place to look for inspiration as we move forward - they've done a lot of things right.

I agree that GHC code is hard to read, and it's an artifact of the fact that it's ~30 years old and add *a lot* of features over time. Not only should we make the code easier to read, but we should modularize better. I feel each pass of the compiler should be in its own package if possible, this will improve compile times as well. The current compile times suck because of the deep module hierachy and because of non-reproducible builds.

**Comment by NickSeagull** [comment]
Yep, this is a big one @jneira , but I wanted to have this here. Didnt understand what self-hoisting supposed to mean 😅 . But still I think it is nice to have this issue to have a "checkmark" for when this is possible.
